src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -34,7 +34,8 @@ public sealed class ReadyToRunCodegenNodeFactory : NodeFactory
             InteropStubManager interopStubManager,
             NameMangler nameMangler,
             VTableSliceProvider vtableSliceProvider,
-            DictionaryLayoutProvider dictionaryLayoutProvider)
+            DictionaryLayoutProvider dictionaryLayoutProvider,
+            EcmaModule inputModule)
```

Comment:
My only gripe with this change is passing a single input module in to the node factory and having the concept of a single module context during compilation. When we eventually do multi-compile where it's A.dll + B.dll + C.dll => Artifact.ni.dll, it will be hard to do if we've set in stone that we have a single compilation input. Maybe it makes more sense to pass in the SignatureContext already built to the NodeFactory constructor so it doesn't have to know about EcmaModules. At some point, we'd modify `SignatureContext` itself to take a list of input assemblies.

Reply:
I completely agree this was a temporary shortcut - I'll probably need to spend some more time thinking how to address this cleanly. AFAIK the current POR is that, even once we enable merging, it will only involve merging of the produced native EXE, the input MSIL assemblies should remain in place and continue being used to resolve metadata tokens and provide IL for JIT when necessary.

In particular, one interesting technical question is whether we'll somehow select a "main" assembly and use it as the carrier for the R2R native image, or whether we'll put the R2R native executable into a completely new file; or perhaps whether we should emit a formal R2R header to each of the MSIL files, somehow pointing all of them to the native image.

These decisions heavily influence how the signature context should work in the merged case. Most notably, I believe it's inevitable that we'll need CoreCLR runtime changes to support the merged case, and the way of making them will affect how the runtime views the "current module" for the purpose of module-relative signature encoding.

I'll try to write up more detailed design thoughts towards the module merging and I'll use this exercise to figure out for myself what exactly needs to happen with the signature context so that it smoothly supports the merged scenario once we implement it.

Reply:
For now I have made the simple change you suggested by passing the SignatureContext to R2R codegen node factory instead of creating it inside it so that we can stop passing around the input file name. I'll follow up separately on finalizing our story for R2R module merging.